### ISSUE TITLE
Provide a page-side approach for `component-pull` & `component-pull-marked`?

### ISSUE THREAD

**Issue by mirisuzanne** [bkardell/shadow-boxing#1]
In a standards world, I would want elements to opt-in (or even opt-in parts?!). But with a polyfill, that adds an extra dependency to each custom element – making them more difficult to share. Polyfill dependencies feel like they should be a page-side decision.

Would it make sense to have an option that is opt-in per-component, but supplied by the page? The page would provide an allow-list for components that should be opted-into `page-push` and `page-push-marked` – creating a sort of post-registration `component-pull`/`component-pull-marked`.

**Comment by bkardell** [comment]
Ah... hmm so do you mean something like... maybe let the `shadow-import` marker take a `DOMTokenList` of tag names it would apply to?
 
```html
<html shadow-style-mode="page-push-marked-allow">
<head>
  <!-- only push these styles down to x-foo and y-bar shadows -->
  <style shadow-import="x-foo y-bar">....</style>
</head>
<body>
   <x-foo>This one</x-foo>
   <y-not>Not this one</y-not>
   <b-cuz>Not this one</b-cuz>
   <i-said>Not this one</i-said>
   <x-bar>This one</x-foo>
</body>
```

**Comment by mirisuzanne** [comment]
Yeah something like that - but maybe also a way to allowlist all page styles, not just one sheet at a time.

**Comment by bkardell** [comment]
I would accept a pull for something here, or if you can help propose how this should actually work (name, where you put it etc) I can prob make this work  - it's really just a filter on what's already there. Seems very reasonable/doable

**Comment by mirisuzanne** [comment]
It might be a bit before I could get to a PR, but roughly I imagine:

- your syntax makes sense for a marked version, mapping specific styles to specific elements
- for the 'opt-into-all-page-styles' approach, I think it could be an attribute on the element itself - maybe even `open-stylable` or something? I think it's getting at the same idea as the open-stylable proposal.
- I don't know at this point if there's a reasonable way to do it per-part instead of the whole element? Like, if you add `part="open-stylable"` to anything internal, styles are imported with a new parent selector like `[part='open-stylable']`? That seems like a stretch case that might be interesting, but maybe less essential. 

I would not expect changes to specificity, but I would expect the global styles to come before shadow styles in the source order. From there, people can use layers to manage priority.

**Comment by bkardell** [comment]
After some offline discussion we're going to focus this issue on the second bullet here (we can open other issues for the other two if we want) and it will be of the 'page-push' philosophy (ie, the components don't have to subclass anything) - we'll call them 'page-push-select' and 'page-push-select-marked'... 


## page-push-select

Pushes all styles from the page into selected shadow roots - those on elements with the `shadow-style-select` attribute.

```
<html shadow-style-mode="page-push-select">
<head>
  <style>....</style>
</head>
<body>
   <x-foo shadow-style-select>This one</x-foo>
   <y-not>Not this one</y-not>
   <b-cuz>Not this one</b-cuz>
   <i-said>Not this one</i-said>
   <x-bar shadow-style-select>This one</x-foo>
</body>
</html>
```

## page-push-select-marked

Pushes styles from the page which are marked with 'shadow-import' into selected shadow roots - those on elements with the ` shadow-style-selected` attribute.

```
<html shadow-style-mode="page-push-select-marked">
<head>
  <!-- only this style will be imported -->
  <style shadow-import>....</style>
  
  <!-- this style will not be imported -->
   <style>....</style>
</head>
<body>
   <x-foo shadow-style-select>This one</x-foo>
   <y-not>Not this one</y-not>
   <b-cuz>Not this one</b-cuz>
   <i-said>Not this one</i-said>
   <x-bar shadow-style-select>This one</x-foo>
</body>
</html>
```


I'll try to work on it this weekend.

**Comment by bkardell** [comment]
@mirisuzanne 

Check these out to see if they represent what you were thinking here before I make any changes... you can fork it or create a pen, or if you want I can add you as a collaborator to that glitch and you can help make it better/fool around with tests, etc

* https://shadow-boxing.glitch.me/index-mia.html  
* https://shadow-boxing.glitch.me/index-mia-marked.html

**Comment by mirisuzanne** [comment]
These look great! Exactly what I had in mind. Thanks!

(well, ok - I didn't really expect yellow text on a lime background…)

**Comment by bkardell** [comment]
closed by #3
